lib/ibandit/iban.rb

```diff
@@ -120,17 +87,57 @@ def valid_check_digits?
       end
     end
 
-    def valid_length?
+    def valid_bank_code_length?
       return unless valid_country_code?
 
-      if iban.size == structure[:total_length]
-        true
-      else
-        @errors[:length] = "Length doesn't match SWIFT specification " \
-                           "(expected #{structure[:total_length]} " \
-                           "characters, received #{iban.size})"
-        false
+      if bank_code.nil?
+        @errors[:bank_code] = 'is required'
+        return false
       end
+
+      return true if bank_code.length == structure[:bank_code_length]
+
+      @errors[:bank_code] = 'is the wrong length: must be ' \
+                            "#{structure[:bank_code_length]}, not " \
+                            "#{bank_code.length}"
+      false
+    end
+
+    def valid_branch_code_length?
+      return unless valid_country_code?
+
+      if branch_code.length == 0 && structure[:branch_code_length] > 0
+        @errors[:branch_code] = 'is required'
+        return false
+      end
+
+      if branch_code.length > 0 && structure[:branch_code_length] == 0
+        @errors[:branch_code] = "is not used in #{country_code}"
+        return false
+      end
+
+      return true if branch_code.length == structure[:branch_code_length]
+
+      @errors[:branch_code] = 'is the wrong length: must be ' \
+                              "#{structure[:branch_code_length]}, not " \
+                              "#{branch_code.length}"
+      false
+    end
+
+    def valid_account_number_length?
+      return unless valid_country_code?
+
+      if account_number.nil?
+        @errors[:account_number] = 'is required'
+        return false
+      end
+
+      return true if account_number.length == structure[:account_number_length]
```

Comment:
I'm guessing the `IBANBuilder` is going to take care of zero-padding. Maybe need to give that a different name, since it transforms params rather than just building an IBAN, or maybe they're separate steps...